runtime/lua/vim/filetype.lua

```diff
@@ -2218,30 +2221,46 @@ local pattern = {
 -- luacheck: pop
 -- luacheck: pop
 
+--- Lookup table/cache for patterns
+--- @alias vim.filetype.pattern_cache { has_env: boolean, has_slash: boolean }
+--- @type table<string,vim.filetype.pattern_cache>
+local pattern_lookup = {}
+
+local function compare_by_priority(a, b)
+  return a[next(a)][2].priority > b[next(b)][2].priority
+end
+
 --- @param t vim.filetype.mapping
 --- @return vim.filetype.mapping[]
+--- @return vim.filetype.mapping[]
 local function sort_by_priority(t)
-  local sorted = {} --- @type vim.filetype.mapping[]
-  for k, v in pairs(t) do
-    local ft = type(v) == 'table' and v[1] or v
+  -- Separate patterns with non-negative and negative priority because they
+  -- will be processed separately
+  local pos = {} --- @type vim.filetype.mapping[]
+  local neg = {} --- @type vim.filetype.mapping[]
+  for pat, maptbl in pairs(t) do
+    local ft = type(maptbl) == 'table' and maptbl[1] or maptbl
     assert(
       type(ft) == 'string' or type(ft) == 'function',
       'Expected string or function for filetype'
     )
 
-    local opts = (type(v) == 'table' and type(v[2]) == 'table') and v[2] or {}
-    if not opts.priority then
-      opts.priority = 0
-    end
-    table.insert(sorted, { [k] = { ft, opts } })
+    -- Parse pattern for common data and cache it once
+    pattern_lookup[pat] = pattern_lookup[pat]
+      or { has_env = pat:find('%$%b{}') ~= nil, has_slash = pat:find('/') ~= nil }
+
+    local opts = (type(maptbl) == 'table' and type(maptbl[2]) == 'table') and maptbl[2] or {}
+    opts.priority = opts.priority or 0
+
+    table.insert(opts.priority >= 0 and pos or neg, { [pat] = { ft, opts } })
   end
-  table.sort(sorted, function(a, b)
-    return a[next(a)][2].priority > b[next(b)][2].priority
-  end)
-  return sorted
+
+  table.sort(pos, compare_by_priority)
+  table.sort(neg, compare_by_priority)
```

Comment:
Light suggestion: we could just sort the built-in table once, then any additions with `vim.filetype.add` can place the new filetypes at the correct position rather than invoking a full sort.

Reply:
That is a reasonable thing to do. I'd leave it for some other time.

Reply:
At least add a todonote for it.

Reply:
It's already approved, so I don't want to force push. Is it OK to leave as is? It would still pop up if/when `vim.filetype.add()` optimization is considered to be done.

Reply:
I'd still prefer it (I don't know where the force-push concern comes from.)

Reply:
OK. I'll add it shortly.

Reply:
Added a TODO comment.